src/raster/types.rs

```diff
@@ -1,41 +1,482 @@
+        let mut is_rounded: libc::c_int = 0;
+
+        let result = unsafe {
+            GDALAdjustValueToDataType(
+                self.gdal_type(),
+                value.into(),
+                &mut is_clamped,
+                &mut is_rounded,
+            )
+        };
+
+        match (is_clamped > 0, is_rounded > 0) {
+            (false, false) => AdjustedValue::Unchanged(result),
+            (true, false) => AdjustedValue::Clamped(result),
+            (false, true) => AdjustedValue::Rounded(result),
+            (true, true) => panic!("Unexpected adjustment result: clamped and rounded."),
+        }
+    }
+
+    /// Determine if converting a value from [`GDALDataType`] described by `self` to one
+    /// described by `other` is potentially lossy.
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::raster::GdalType;
+    /// assert!(<i16>::descriptor().is_conversion_lossy(<u8>::descriptor()))
+    /// ```
+    pub fn is_conversion_lossy(&self, other: Self) -> bool {
+        let r = unsafe { GDALDataTypeIsConversionLossy(self.gdal_type(), other.gdal_type()) };
+        r != 0
+    }
+
+    /// Subset of the GDAL data types supported by Rust bindings.
+    pub fn available_types() -> &'static [GdalTypeDescriptor] {
+        use GDALDataType::*;
+        &[
+            GdalTypeDescriptor(GDT_Byte),
+            GdalTypeDescriptor(GDT_UInt16),
+            GdalTypeDescriptor(GDT_Int16),
+            GdalTypeDescriptor(GDT_UInt32),
+            GdalTypeDescriptor(GDT_Int32),
+            #[cfg(all(major_ge_3, minor_ge_5))]
+            GdalTypeDescriptor(GDT_UInt64),
+            #[cfg(all(major_ge_3, minor_ge_5))]
+            GdalTypeDescriptor(GDT_Int64),
+            GdalTypeDescriptor(GDT_Float32),
+            GdalTypeDescriptor(GDT_Float64),
+        ]
+    }
+}
+
+impl Debug for GdalTypeDescriptor {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("GdalTypeDescriptor")
+            .field("name", &self.name())
+            .field("bits", &self.bits())
+            .field("signed", &self.is_signed())
+            .field("floating", &self.is_floating())
+            .field("gdal_ordinal", &self.gdal_type())
+            .finish()
+    }
+}
+
+impl Display for GdalTypeDescriptor {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.name())
+    }
+}
+
+/// Converts from a possible [`GDALDataType`] ordinal value to a [`GdalTypeDescriptor`].
+///
+/// # Example
+///
+/// ```rust, no_run
+/// use gdal::raster::{GdalType, GdalTypeDescriptor};
+/// let gdt: GdalTypeDescriptor = 3.try_into().unwrap();
+/// println!("{gdt:#?}")
+/// ```
+impl TryFrom<GDALDataType::Type> for GdalTypeDescriptor {
+    type Error = GdalError;
+
+    fn try_from(value: GDALDataType::Type) -> std::result::Result<Self, Self::Error> {
+        let wrapped = GdalTypeDescriptor(value);
+        if !GdalTypeDescriptor::available_types().contains(&wrapped) {
```

Comment:
this could be a match

Reply:
@jdroenner I read up on slice patterns, and I still couldn't figure out how to implement `contains` with a `match` 😰